### PR TITLE
Markdown editor in the Project Lab

### PR DESCRIPTION
**Pull request by edpaget**
Makes a universal editor for markdown.

### CODE REVIEW

File: app/components/markdown-editor.cjsx [hunk 1]
```diff
@@ -17,32 +19,125 @@ module.exports = React.createClass
...
+  wrapSelectionIn: (wrapFn, opts = {}) ->
+    # helper to call markdown-insert functions on the textarea
+    # wrapFn takes / returns a string (from ./lib/markdown-insert.cjsx)
+
+    textarea = @refs.textarea.getDOMNode()
+    selection = m.getSelection(textarea)
+    {text, cursor} = wrapFn(selection)
+
+    m.insertAtCursor(text, textarea, cursor, opts)
+    @onInputChange()
```
Comment (brian-c) [inline app/components/markdown-editor.cjsx]: This method doesn't exist. Also on line 143.

**Comment by brian-c** [comment]
I haven't dug into it too much, but it looks like maybe `@props.onChange` isn't being called after using the formatting buttons. I select some text, hit the Bold button, then the Preview button, and I lose the boldness.

The formatting buttons also seem to scroll the text down.

**Comment by edpaget** [comment]
I think I've taken care of that now. It looks like you were right about onChange not firing when you used the buttons to insert formatting.

**Comment by edpaget** [comment]
@aweiksnar I don't know if you also want to look at this since it touches a bunch of talk stuff.

**Comment by aweiksnar** [comment]
Yeah I'll take a peek and make some test comments

File: app/components/markdown-editor.cjsx (hunk 1 above)
Comment (edpaget) [inline app/components/markdown-editor.cjsx]: Added it back in to handle the onChange events.

**Comment by aweiksnar** [comment]
Only thing I've noticed that isn't working for me is clicking bold/italic/etc submits the comment in talk

File: app/talk/comment-box.cjsx [hunk 2]
```diff
@@ -44,68 +43,31 @@ module?.exports = React.createClass
...
         @setFeedback @props.submitFeedback
       .catch (e) =>
         @setState(error: e.message, loading: false)
 
-  onPreviewClick: (e) ->
-    @toggleComponent('preview')
-
-  onHelpClick: (e) ->
-    @toggleComponent('help')
+  onInputChange: (e) ->
```
Comment (aweiksnar) [inline app/talk/comment-box.cjsx]: Calling e.preventDefault() here remedies that comment submitting issue

**Comment by edpaget** [comment]
Ugh fix one thing create another problem.

File: app/talk/comment-box.cjsx (hunk 2 above)
Comment (edpaget) [inline app/talk/comment-box.cjsx]: It seems like some button is acting as `type=submit` since that shouldn't be necessary.

Comment (edpaget) [inline app/talk/comment-box.cjsx]: Yeah none of those buttons had a type declared. I added it and it fixed that problem for me.

Comment (aweiksnar) [inline app/talk/comment-box.cjsx]: Ah nice!

File: app/talk/comment-box.cjsx [hunk 3]
```diff
@@ -159,7 +88,8 @@ module?.exports = React.createClass
             <CommentPreview header={null} content={@state.reply}/>
           </div>}
 
-        <textarea value={@state.content} onChange={@onInputChange} ref="textarea" placeholder={@props.placeholder} />
+        <MarkdownEditor className="full" value={@state.content} onChange={@onInputChange} helpText={<CommentHelp />}/>
```
Comment (aweiksnar) [inline app/talk/comment-box.cjsx]: Can you add `placeholder={@props.placeholder}` back in here?

**Comment by aweiksnar** [comment]
Cool LGT:shipit: by me

**Comment by brian-c** [comment]
The only thing still stuck for me is the scrolling. I've got it fixed in `insertAtCursor` in markdown-insert.coffee by storing `{scrollLeft, scrollTop} = input` at the start and resetting them at the end, but I'm not sure why it's jumping to the end to begin with.

**Comment by edpaget** [comment]
@brian-c it looks like this is an old bug in Chrome/Webkit. https://code.google.com/p/chromium/issues/detail?id=75072. I've worked around it by reading the `scrollTop` of the input before the focus event and reseting it after focus is called.

**Pull request merged**